asm/exceptasm.s

```diff
@@ -69,646 +81,897 @@ glabel __osIntTable
-/* 004EBC 800042BC 001BD840 */  sll   $k1, $k1, 1
-/* 004EC0 800042C0 3C1A8001 */  lui   $k0, %hi(__osRcpImTable) # $k0, 0x8001
-/* 004EC4 800042C4 275A2160 */  addiu $k0, %lo(__osRcpImTable) # addiu $k0, $k0, 0x2160
-/* 004EC8 800042C8 037AD821 */  addu  $k1, $k1, $k0
-/* 004ECC 800042CC 977B0000 */  lhu   $k1, ($k1)
-/* 004ED0 800042D0 3C1AA430 */  lui   $k0, %hi(D_A430000C) # $k0, 0xa430
-/* 004ED4 800042D4 275A000C */  addiu $k0, %lo(D_A430000C) # addiu $k0, $k0, 0xc
-/* 004ED8 800042D8 AF5B0000 */  sw    $k1, ($k0)
-/* 004EDC 800042DC 00000000 */  nop   
-/* 004EE0 800042E0 00000000 */  nop   
-/* 004EE4 800042E4 00000000 */  nop   
-/* 004EE8 800042E8 00000000 */  nop   
-/* 004EEC 800042EC 42000018 */  eret  
-glabel __osCleanupThread
-/* 004EF0 800042F0 0C0010D0 */  jal   osDestroyThread
-/* 004EF4 800042F4 00002025 */   move  $a0, $zero
+    li      $at, CAUSE_CEMASK
+    and     $t1, $t0, $at
+    srl     $t1, $t1, CAUSE_CESHIFT
+    li      $t2, 1 # if not coprocessor 1, panic
+    bne     $t1, $t2, panic
+     nop
+    # Mark cop1 as usable for previous thread
+    lw      $k1, THREAD_SR($k0)
+    li      $at, SR_CU1
+    li      $t1, 1
+    or      $k1, $k1, $at
+    sw      $t1, THREAD_FP($k0)
+    b       enqueueRunning
+     sw     $k1, THREAD_SR($k0)
+END(__osException)
+
+/**
+ *  void __osEnqueueAndYield(OSThread** threadQueue);
+ *
+ *  Voluntary thread yielding.
+ *  Enqueues the currently running thread to the top of the
+ *   thread queue `threadQueue` and yields to the highest priority
+ *   unblocked runnable thread.
+ */
+LEAF(__osEnqueueAndYield)
+    lui     $a1, %hi(__osRunningThread)
+    lw      $a1, %lo(__osRunningThread)($a1)
+    # Save SR
+    mfc0    $t0, C0_SR
+    lw      $k1, THREAD_FP($a1)
+    ori     $t0, $t0, SR_EXL
+    sw      $t0, THREAD_SR($a1)
+    # Save callee-saved registers
+    sd      $s0, THREAD_S0($a1)
+    sd      $s1, THREAD_S1($a1)
+    sd      $s2, THREAD_S2($a1)
+    sd      $s3, THREAD_S3($a1)
+    sd      $s4, THREAD_S4($a1)
+    sd      $s5, THREAD_S5($a1)
+    sd      $s6, THREAD_S6($a1)
+    sd      $s7, THREAD_S7($a1)
+    sd      $gp, THREAD_GP($a1)
+    sd      $sp, THREAD_SP($a1)
+    sd      $fp, THREAD_S8($a1)
+    sd      $ra, THREAD_RA($a1)
+    # Save FPU callee-saved registers if the current thread has used the FPU
+    beqz    $k1, 1f
+     sw     $ra, THREAD_PC($a1)
+    cfc1    $k1, C1_FPCSR
+    sdc1    $f20, THREAD_FP20($a1)
+    sdc1    $f22, THREAD_FP22($a1)
+    sdc1    $f24, THREAD_FP24($a1)
+    sdc1    $f26, THREAD_FP26($a1)
+    sdc1    $f28, THREAD_FP28($a1)
+    sdc1    $f30, THREAD_FP30($a1)
+    sw      $k1, THREAD_FPCSR($a1)
+1:
+    lw      $k1, THREAD_SR($a1)
+    andi    $t1, $k1, SR_IMASK
+    beqz    $t1, 2f
+     nop
+    # If any CPU interrupts are enabled in SR, mask previous thread's SR
+    # with the CPU bits of the global interrupt mask
+    lui     $t0, %hi(__OSGlobalIntMask)
+    addiu   $t0, %lo(__OSGlobalIntMask)
```

Comment:
I've been avoiding using instruction macros that expand to multiple instructions since they can't be used consistently. If a working `as0` recomp ever happens I'd like to redo the asm files to have this sort of thing since we'd be able to do it properly there.